torch/fx/experimental/symbolic_shapes.py

```diff
@@ -800,29 +800,73 @@ def render(self, source: Source):
 DimConstraint = Union[StrictMinMaxConstraint, RelaxedUnspecConstraint, None]
 
 @dataclass(frozen=True)
-class EqualityConstraint(Constraint):
+class EqualityConstraints(Constraint):
     """
-    Given pairs of sources corresponding to pairs of dynamic dimensions that
-    are specified equal, represent them in a union-find data structure so that
-    we can efficiently check whether two such sources are transitively equal.
+    Represent and decide various kinds of equality constraints between input sources.
+
+    A "source pair" is a pair of input sources for dynamic dimensions that
+    are specified equal. We represent `source_pairs` in a union-find forest
+    so that we can efficiently check whether two such sources are transitively equal.
+
+    A "derived equality" relates an input source to an expression over a root.
+    The root can be another input source, corresponding to some dynamic dimension,
+    or a phantom symbol that does not directly represent any dynamic dimension. We
+    represent `derived_equalities` involving input sources in a transitively-closed map
+    so that we can efficiently check whether an input source is transitively equal to
+    a given expression over another input source.
+    (NOTE: In contrast, it is easy to decide whether an input source is transitively equal
+    to a given expression over a phantom symbol; such expressions are already in canonical
+    form and so the problem reduces to symbolic expression equality.)
```

Comment:
An example feels like it would help here. Maybe even the counterexample you gave earlier.

```
Suppose we have [a = b, b = c], [d = c + 1, e = d - 1]

We first construct a union find with source_pairs, giving us

_parents = {a: a, b: a, c: a}

Then, we compute canonical symbolic expressions, recursively applying unary expressions until we bottom out

_defs = {d: c + 1, e: (c + 1) - 1 aka c}
```

Reply:
Good idea.